sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/specialized/BlockBlobAsyncApiTests.java

```diff
@@ -0,0 +1,2658 @@
+                                                byte[] expectedDownload) {
+        StepVerifier.create(blobAsyncClient.upload(Flux.fromIterable(Arrays.asList(buffer1, buffer2, buffer3)),
+            null, true)).assertNext(it -> assertNotNull(it.getETag())).verifyComplete();
+
+        StepVerifier.create(FluxUtil.collectBytesInByteBufferStream(blobAsyncClient.downloadStream()))
+            .assertNext(it -> TestUtils.assertArraysEqual(it, expectedDownload))
+            .verifyComplete();
+    }
+
+    private static Stream<Arguments> asyncBufferedUploadEmptyBuffersSupplier() {
+        return Stream.of(Arguments.of(ByteBuffer.wrap("Hello".getBytes(StandardCharsets.UTF_8)),
+                ByteBuffer.wrap(" ".getBytes(StandardCharsets.UTF_8)),
+                ByteBuffer.wrap("world!".getBytes(StandardCharsets.UTF_8)),
+                "Hello world!".getBytes(StandardCharsets.UTF_8)),
+            Arguments.of(ByteBuffer.wrap("Hello".getBytes(StandardCharsets.UTF_8)),
+                ByteBuffer.wrap(" ".getBytes(StandardCharsets.UTF_8)),
+                ByteBuffer.wrap(new byte[0]),
+                "Hello ".getBytes(StandardCharsets.UTF_8)),
+            Arguments.of(ByteBuffer.wrap("Hello".getBytes(StandardCharsets.UTF_8)),
+                ByteBuffer.wrap(new byte[0]),
+                ByteBuffer.wrap("world!".getBytes(StandardCharsets.UTF_8)),
+                "Helloworld!".getBytes(StandardCharsets.UTF_8)),
+            Arguments.of(ByteBuffer.wrap(new byte[0]),
+                ByteBuffer.wrap(" ".getBytes(StandardCharsets.UTF_8)),
+                ByteBuffer.wrap("world!".getBytes(StandardCharsets.UTF_8)),
+                " world!".getBytes(StandardCharsets.UTF_8)));
+    }
+
+    // Only run these tests in live mode as they use variables that can't be captured.
+    @ParameterizedTest
+    @MethodSource("asyncBufferedUploadSupplier")
+    @EnabledIf("com.azure.storage.blob.BlobTestBase#isLiveMode")
+    public void asyncBufferedUpload(int dataSize, long bufferSize, int numBuffs, int blockCount) {
+        BlobAsyncClient asyncClient = getPrimaryServiceClientForWrites(bufferSize)
+            .getBlobContainerAsyncClient(blobAsyncClient.getContainerName())
+            .getBlobAsyncClient(blobAsyncClient.getBlobName());
+
+        ByteBuffer data = getRandomData(dataSize);
+        ParallelTransferOptions parallelTransferOptions = new ParallelTransferOptions()
+            .setBlockSizeLong(bufferSize).setMaxConcurrency(numBuffs).setMaxSingleUploadSizeLong(4L * Constants.MB);
+        asyncClient.upload(Flux.just(data), parallelTransferOptions, true).block();
+        data.position(0);
+
+        // Due to memory issues, this check only runs on small to medium sized data sets.
+        if (dataSize < 100 * 1024 * 1024) {
+            StepVerifier.create(collectBytesInBuffer(blockBlobAsyncClient.downloadStream()))
+                .assertNext(it -> assertEquals(it, data))
+                .verifyComplete();
+        }
```

Comment:
In the future we should think about redesigning this test to allow comparison of huge blobs. Instead of having to load everything into memory we can compare iteratively.